File: detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/Application.java
package io.buildfoundation.bazel.detekt;

import io.buildfoundation.bazel.detekt.execute.Executable;
import io.buildfoundation.bazel.detekt.execute.ExecutableResult;
import io.buildfoundation.bazel.detekt.execute.WorkerExecutable;
import io.buildfoundation.bazel.detekt.stream.Streams;
import io.buildfoundation.bazel.detekt.stream.WorkerStreams;
import io.reactivex.rxjava3.core.Scheduler;

public interface Application {

    void run(String[] args);

    final class OneShot implements Application {

        private final Executable executable;
        private final Streams streams;
        private final Platform platform;

        OneShot(Executable executable, Streams streams, Platform platform) {
            this.executable = executable;
            this.streams = streams;
            this.platform = platform;
        }

        @Override
        public void run(String[] args) {
            ExecutableResult result = executable.execute(args);

            if (result instanceof ExecutableResult.Failure) {
                streams.error().println(result.output());
            }

            platform.exit(result.statusCode());
        }
    }

    final class Worker implements Application {

        private final WorkerExecutable executable;
        private final WorkerStreams streams;
        private final Scheduler scheduler;

        Worker(WorkerExecutable executable, WorkerStreams streams, Scheduler scheduler) {
            this.executable = executable;
            this.streams = streams;
            this.scheduler = scheduler;
        }

        @Override
        public void run(String[] args) {
            streams.request()
                .subscribeOn(scheduler)
                .map(executable::execute)
                .blockingSubscribe(streams.response());
        }
    }
}

File: detekt/wrapper/src/main/kotlin/io/buildfoundation/bazel/detekt/Application.kt
package io.buildfoundation.bazel.detekt

import io.buildfoundation.bazel.detekt.execute.Executable
import io.buildfoundation.bazel.detekt.execute.Result
import io.buildfoundation.bazel.detekt.execute.WorkerExecutable
import io.buildfoundation.bazel.detekt.stream.Streams
import io.buildfoundation.bazel.detekt.stream.WorkerStreams
import io.reactivex.Scheduler

internal interface Application {

    fun run(args: Array<String>)

    class OneShot(
            private val executable: Executable,
            private val streams: Streams,
            private val platform: Platform
    ) : Application {

        override fun run(args: Array<String>) {
            val result = executable.execute(args)

            if (result is Result.Failure) {
                streams.error.println(result.description)
            }

            platform.exit(result.consoleStatusCode)
        }
    }

    class Worker(
            private val executable: WorkerExecutable,
            private val streams: WorkerStreams,
            private val scheduler: Scheduler
    ) : Application {

        override fun run(args: Array<String>) {
            streams.request
                    .subscribeOn(scheduler)
                    .map { executable.execute(it) }
                    .blockingSubscribe(streams.response)
        }
    }
}